### DISCUSSION TITLE
Deprecation notice for canContainersAccessInstanceRole prop in ECS construct

### DISCUSSION THREAD

**Discussion by QuantumNeuralCoder** [aws/aws-cdk#32609] [General]
Hi All,

This is to inform you of an upcoming change in support of ```canContainersAccessInstanceRole``` property used in the ECS construct.

AWS Cloud Development Kit (AWS CDK) will deprecate the ```canContainersAccessInstanceRole``` property used in the ECS Cluster.AddCapacity [[1](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AddCapacityOptions.html#cancontainersaccessinstancerole )], Cluster.addAutoScalingGroupCapacity [[2](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AddAutoScalingGroupCapacityOptions.html#cancontainersaccessinstancerole )] APIs, and AsgCapacityProvider [[3](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AsgCapacityProvider.html#cancontainersaccessinstancerole)] construct. This feature will be removed from the CDK codebase by December 12, 2025.

**Rationale for Deprecation** 
We are deprecating this feature because of challenges in supporting the security sensitive nature of this flag going forward. CDK cannot guarantee the correct execution of ```canContainersAccessInstanceRole``` across all platforms that depend on platform images, container engines, and future updates to these components.

**Recommended Approach**
We recommend you use the ECS service's recommended configurations to block IMDS on EC2 instances moving forward:

For Windows and Linux, follow the guidance in the [AWS documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-metadata-limiting-access.html).

For Bottlerocket, refer to the [discussion on the Bottlerocket GitHub repository](https://github.com/bottlerocket-os/bottlerocket/discussions/1760#discussioncomment-1392343).

**Disabling the Deprecated Feature**
To ensure this deprecated feature is not used in your environments, we strongly suggest enabling the ```Disable_ECS_IMDS_Blocking``` feature flag. This flag will be automatically enabled for all new CDK projects by January 8, 2025, created using the _cdk init_ command, but will be disabled for your existing projects. If this flag is enabled, the CDK synthesis and deploy commands will fail whenever ```canContainersAccessInstanceRole``` is set to false. This has been done to signal that the instance role access from containers should be blocked using the recommended approaches. 

**Temporary Usage of the Deprecated Feature**
If for any reason, you need to retain usage of ```canContainersAccessInstanceRole``` to block container access to IMDS on Linux, you can set the feature flag value ```Disable_ECS_IMDS_Blocking``` to false to avoid failures on synthesis and deploy commands after the launch of this feature flag by January 8, 2025.

Please note that the CDK team does not recommend this, and you should only use it as a short-term solution. You will also continue to see warning messages when using this deprecated feature, as there is no guarantee that the current approach will keep working for any future updates to the Linux kernel or container engines running on them. 

CDK does not support this feature for Windows and Bottlerocket images. CDK will raise warning messages for these operating systems until the final removal of this feature entirely by December 12, 2025.

**Provide Feedback**
We encourage you to share your feedback on this change on this discussion on Github.

[1] https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AddCapacityOptions.html#cancontainersaccessinstancerole 
[2] https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AddAutoScalingGroupCapacityOptions.html#cancontainersaccessinstancerole 
[3] https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.AsgCapacityProvider.html#cancontainersaccessinstancerole

**Comment by valkarinc** [comment]
Nice, with the upcoming deprecation of the `canContainersAccessInstanceRole `property in the ECS construct, it's clear that AWS is focusing on enhancing security and maintaining up-to-date practices. The new methods to block IMDS on EC2 instances for Windows, Linux, and Bottlerocket are straightforward, and I'll start implementing them to ensure my setups remain secure. The `Disable_ECS_IMDS_Blocking` feature flag seems like a useful tool during this transition. Thanks for the clear instructions and advance notice; they'll help me adjust to these changes smoothly.

Appreciate it!

**Comment by pwrmiller** [comment]
I'm convinced that this is a regression, not an improvement to security posture. It would seem that AWS no longer wants to support their own best practices for deploying ECS clusters. I understand why this might be being considered - namely, the complexity of the aws_cdk.aws_ecs constructs specifically around capacity providers, mixed instances polices and more might mean that this flag isn't properly set. 

As a note, this flag was so rushed that there's even a typo in its implementation:
```
[Warning at /us/Application/ECS/Default] Blocking container access to instance role will be deprecated. Use the @aws-cdk/aws-ecs:enableImdsBlockingDeprecatedFeature feature flagto keep this feature temporarily. See https://github.com/aws/aws-cdk/discussions/32609 [ack: @aws-cdk/aws-ecs:deprecatedImdsBlocking]
```

"flagto" is the typo. 

As a case in point, associating a launch template with a mixed instances policy rather than a cluster will trigger these warnings, but associating a launch template with a cluster will not. There's nothing the customer can do to work around the feature in this particular case without using significant, complicated escape hatches.

**Comment by patchwork01** [comment]
Reading between the lines from this announcement and the documentation, although it's not directly stated, it seems likely that what this flag did was to add an iptables rule to prevent requests from the "apache" user reaching IMDS.

That's very specific to a case where someone is running a web server on the "apache" user of the EC2, and has code in the server that makes web calls on behalf of external actors. I would think anyone would expect to need to lock that down themselves somehow, and I'd be worried if someone expected AWS to make that safe for them with a change to the CDK.

If that's what happened, it makes perfect sense that this has been removed. It's disappointing that it's still not clear what this flag does, and as a result it's unclear whether removing it on an existing application will have security implications.

Please can we have some clarification of what this flag did? Is what I described above more or less accurate?

**Reply by ab** [reply]
Here is the code for that param before it was removed: https://github.com/aws/aws-cdk/blob/7395b41687718366940bddb9117bc3a8e9f09863%5E/packages/aws-cdk-lib/aws-ecs/lib/cluster.ts#L791-L810

The flag causes a few things to be added to the EC2 user-data:

Looks like it added an iptables rule:
```sh
iptables --insert FORWARD 1 --in-interface docker+ --destination 169.254.169.254/32 --jump DROP
service iptables save
```

And also this:
```sh
echo ECS_AWSVPC_BLOCK_IMDS=true >> /etc/ecs/ecs.config
```

It seems like there were issues moving from AL2 to AL2023 because of the iptables stuff (AL2023 uses nftables), leading to the decision to deprecate and remove the `canContainersAccessInstanceRole` parameter.

https://github.com/aws/aws-cdk/issues/32469
https://github.com/aws/aws-cdk/issues/28518
